conclave-host/src/main/kotlin/com/r3/conclave/host/internal/GramineEnclaveHandle.kt

```diff
@@ -1,89 +1,91 @@
         copyGramineFilesToWorkingDirectory()
+
+        /**
+         * Not all threads in the enclave are necessarily the result of conclave calls.
+         * To minimise the likelihood of deadlocks, we don't allow conclave to use all available threads.
+         */
+        val maxConcurrentCalls = getEnclaveThreadCountFromManifest() / 2
+        check(maxConcurrentCalls > 0)
+
+        /** Create a socket host interface, let the system allocate the port. */
+        enclaveInterface = SocketHostEnclaveInterface(0, maxConcurrentCalls)
     }
 
     companion object {
         const val GRAMINE_ENCLAVE_JAR_NAME = "enclave-shadow.jar"
         const val GRAMINE_ENCLAVE_MANIFEST = "java.manifest"
 
-        fun getDummyAttestation(): EnclaveInstanceInfoImpl {
-            val signingKeyPair = SignatureSchemeEdDSA().generateKeyPair()
-            val encryptionPrivateKey = Curve25519PrivateKey.random()
-            val measurement = SHA256Hash.wrap(Random.nextBytes(32))
-            val cpuSvn = OpaqueBytes(Random.nextBytes(16))
-            val mrsigner = SHA256Hash.wrap(Random.nextBytes(32))
-            val isvProdId = 1
-            val isvSvn = 1
-            val reportBody = Cursor.allocate(SgxReportBody).apply {
-                this[SgxReportBody.cpuSvn] = cpuSvn.buffer()
-                this[SgxReportBody.mrenclave] = measurement.buffer()
-                this[SgxReportBody.mrsigner] = mrsigner.buffer()
-                this[SgxReportBody.isvProdId] = isvProdId
-                this[SgxReportBody.isvSvn] = isvSvn
-                this[SgxReportBody.reportData] =
-                    SHA512Hash.hash(signingKeyPair.public.encoded + encryptionPrivateKey.publicKey.encoded).buffer()
-            }
-            val timestamp = Instant.now()
-
-            return EnclaveInstanceInfoImpl(signingKeyPair.public, encryptionPrivateKey.publicKey,
-                MockAttestation(timestamp, reportBody.asReadOnly(), false))
-        }
-
         private val logger = loggerFor<GramineEnclaveHandle>()
     }
 
     override fun initialise() {
-        processGramineDirect = ProcessBuilder()
-            .inheritIO()
-            .directory(enclaveDirectory.toFile())
-            .command("gramine-direct", "java", "-cp", GRAMINE_ENCLAVE_JAR_NAME, "com.r3.conclave.enclave.internal.GramineEntryPoint")
-            .start()
+        /** Start the local call interface. */
+        val interfaceStartTask = FutureTask { enclaveInterface.start() }
+        val interfaceStartThread = Thread(interfaceStartTask).apply { start() }
+
+        try {
+            /** Start the enclave process, passing the port that the call interface is listening on. */
+            processGramineDirect = ProcessBuilder()
+                .inheritIO()
+                .directory(enclaveDirectory.toFile())
+                .command("gramine-direct", "java", "-cp", GRAMINE_ENCLAVE_JAR_NAME, "com.r3.conclave.enclave.internal.GramineEntryPoint", enclaveInterface.port.toString())
```

Comment:
I don't think having that line in the manifest is necessarily bad, provided the code that handles the command line parameters (`GrameneEntryPoint`) is safe. Also, any form of inter-process communication is going to require that the enclave know how to address the host or vice versa anyway, so I'm not sure there is a way to avoid having to pass _something_ in an un-trusted manner be it by file or by using command line parameters.

Reply:
My point is "do we want to make the enclave-side port part of the enclave trusted codebase"? Not sure, I will think about this. I feel that there might be some risk if we allow the same enclave to run the same secure enclave on different ports for example, but can't think of a specific case right now.
Also, as things stand, only the `java` command is actually signed to trust under Gramine, i.e. attackers can really run whatever the `java` command allows them to do under Gramine, for whatever reason.
It is safer if even the `-cp JARNAME ...` part is covered by the signing step and this is possible by using the suggested mechanisms in the docs
```
loader.argv = ["arg0", "arg1", "arg2", ...]

or

loader.argv_src_file = "file:file_with_serialized_argv"
.
```

Finally, I would not really leave a line of the manifest where it says `__insecure`. We will need to explain and discuss and I expect this to be annoying. Of course, if this is our only possibility we will need to provide plenty of explanations

Reply:
I've added a TODO regarding sensitisation of the command line parameters.